File: backend/src/main/java/com/example/finance/service/TransactionService.java
package com.example.finance.service;

import com.example.finance.model.Transaction;
import com.example.finance.model.Transaction.TransactionType;
import com.example.finance.repository.TransactionRepository;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.stereotype.Service;

import java.math.BigDecimal;
import java.time.LocalDateTime;
import java.util.List;
import java.util.Optional;

@Service
public class TransactionService {

    private final TransactionRepository transactionRepository;

    @Autowired
    public TransactionService(TransactionRepository transactionRepository) {
        this.transactionRepository = transactionRepository;
    }

    // Create a new transaction
    public Transaction createTransaction(Transaction transaction) {
        return transactionRepository.save(transaction);
    }

    // Get all transactions
    public List<Transaction> getAllTransactions() {
        return transactionRepository.findAll();
    }

    // Get transaction by ID
    public Optional<Transaction> getTransactionById(Long id) {
        return transactionRepository.findById(id);
    }

    // Update an existing transaction
    public Transaction updateTransaction(Transaction transaction) {
        return transactionRepository.save(transaction);
    }

    // Delete a transaction
    public void deleteTransaction(Long id) {
        transactionRepository.deleteById(id);
    }

    // Get transactions by type (INCOME or EXPENSE)
    public List<Transaction> getTransactionsByType(TransactionType type) {
        return transactionRepository.findByType(type);
    }

    // Get transactions by category
    public List<Transaction> getTransactionsByCategory(String category) {
        return transactionRepository.findByCategory(category);
    }

    // Get transactions between two dates
    public List<Transaction> getTransactionsBetweenDates(LocalDateTime startDate, LocalDateTime endDate) {
        return transactionRepository.findByDateBetween(startDate, endDate);
    }

    // Get transactions by description (partial match)
    public List<Transaction> getTransactionsByDescription(String description) {
        return transactionRepository.findByDescriptionContainingIgnoreCase(description);
    }

    // Calculate total income
    public BigDecimal calculateTotalIncome() {
        BigDecimal totalIncome = transactionRepository.calculateTotalIncome();
        return totalIncome != null ? totalIncome : BigDecimal.ZERO;
    }

    // Calculate total expenses
    public BigDecimal calculateTotalExpenses() {
        BigDecimal totalExpenses = transactionRepository.calculateTotalExpenses();
        return totalExpenses != null ? totalExpenses : BigDecimal.ZERO;
    }

    // Calculate current balance
    public BigDecimal calculateBalance() {
        BigDecimal balance = transactionRepository.calculateBalance();
        return balance != null ? balance : BigDecimal.ZERO;
    }
}

File: frontend/src/app/services/transaction.service.ts
import { Injectable } from '@angular/core';
import { HttpClient } from '@angular/common/http';
import { Observable } from 'rxjs';
import { Transaction, TransactionType } from '../models/transaction.model';

@Injectable({
  providedIn: 'root'
})
export class TransactionService {
  private apiUrl = 'http://localhost:8080/api/transactions';

  constructor(private http: HttpClient) { }

  // Get all transactions
  getAllTransactions(): Observable<Transaction[]> {
    return this.http.get<Transaction[]>(this.apiUrl);
  }

  // Get transaction by ID
  getTransactionById(id: number): Observable<Transaction> {
    return this.http.get<Transaction>(`${this.apiUrl}/${id}`);
  }

  // Create a new transaction
  createTransaction(transaction: Transaction): Observable<Transaction> {
    return this.http.post<Transaction>(this.apiUrl, transaction);
  }

  // Update an existing transaction
  updateTransaction(transaction: Transaction): Observable<Transaction> {
    return this.http.put<Transaction>(`${this.apiUrl}/${transaction.id}`, transaction);
  }

  // Delete a transaction
  deleteTransaction(id: number): Observable<void> {
    return this.http.delete<void>(`${this.apiUrl}/${id}`);
  }

  // Get transactions by type (INCOME or EXPENSE)
  getTransactionsByType(type: TransactionType): Observable<Transaction[]> {
    return this.http.get<Transaction[]>(`${this.apiUrl}/type/${type}`);
  }

  // Get transactions by category
  getTransactionsByCategory(category: string): Observable<Transaction[]> {
    return this.http.get<Transaction[]>(`${this.apiUrl}/category/${category}`);
  }

  // Get transactions between two dates
  getTransactionsBetweenDates(startDate: string, endDate: string): Observable<Transaction[]> {
    return this.http.get<Transaction[]>(`${this.apiUrl}/date-range?startDate=${startDate}&endDate=${endDate}`);
  }

  // Get transactions by description (partial match)
  getTransactionsByDescription(description: string): Observable<Transaction[]> {
    return this.http.get<Transaction[]>(`${this.apiUrl}/search?description=${description}`);
  }

  // Get financial summary (income, expenses, balance)
  getFinancialSummary(): Observable<{totalIncome: number, totalExpenses: number, balance: number}> {
    return this.http.get<{totalIncome: number, totalExpenses: number, balance: number}>(`${this.apiUrl}/summary`);
  }
}
